File: Data Structures & Algorithms/partition-equal-subset-sum/submission-3.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        total = sum(nums)
        target = total // 2
        if total % 2:
            return False

        dp = {}

        def dfs(i,t):
            if i >= len(nums):
                return False
            if t == 0:
                return True

            if (i,t) in dp:
                return dp[(i,t)]

            dp[(i,t)] = dfs(i+1,t)

            if t - nums[i] >= 0:
                dp[(i,t)] = dp[(i,t)] or dfs(i+1, t-nums[i])

            return dp[(i,t)]
        return dfs(0,target)

File: Data Structures & Algorithms/perfect-squares/submission-3.rs
impl Solution {
    pub fn num_squares(n: i32) -> i32 {
        Self::dfs(0, n, &mut vec![-1; n as usize])
    }

    fn dfs(sum: i32, n: i32, memo: &mut Vec<i32>) -> i32 {
        if sum == n {
            return 0;
        }
        if memo[sum as usize] != -1 {
            return memo[sum as usize];
        }
        let mut i = 1;
        let mut res = i32::MAX;
        while i*i <= n - sum {
            res = min(res, 1 + Self::dfs(sum+i*i, n, memo));
            i += 1;
        }
        memo[sum as usize] = res;
        res
    }
}
